pkg/vere/mdns.c

```diff
@@ -0,0 +1,220 @@
+  mdns_cb*      cb;
+  DNSServiceRef sref;
+  char*         who;
+  char*         our;
+  uint16_t      port;
+  uv_poll_t*    poll;
+  void*         context;
+} mdns_payload;
+
+
+static void getaddrinfo_cb(uv_getaddrinfo_t* req, int status, struct addrinfo* res) {
+  mdns_payload* payload = (mdns_payload*)req->data;
+
+  if (status < 0) {
+    u3l_log("mdns: getaddrinfo error: %s", uv_strerror(status));
+    DNSServiceRefDeallocate(payload->sref);
+    c3_free(payload);
+    c3_free(req);
+    uv_freeaddrinfo(res);
+    return;
+  }
+
+  struct sockaddr_in* addr = (struct sockaddr_in*)res->ai_addr;
+  payload->cb(payload->who, addr->sin_addr.s_addr, payload->port, payload->context);
+
+  c3_free(payload->who);
+  uv_freeaddrinfo(res);
+  c3_free(req);
+  DNSServiceRefDeallocate(payload->sref);
+  c3_free(payload);
+
+}
+
+static void resolve_cb(DNSServiceRef sref,
+                       DNSServiceFlags f,
+                       uint32_t interface,
+                       DNSServiceErrorType err,
+                       const char *name,
+                       const char *host,
+                       uint16_t port,
+                       uint16_t tl,
+                       const unsigned char *t,
+                       void *context)
+{
+  mdns_payload* payload = (mdns_payload*)context;
+
+  uv_poll_stop(payload->poll);
+  c3_free(payload->poll);
+
+  if (err != kDNSServiceErr_NoError) {
+    u3l_log("mdns: dns resolve error %d", err);
+    c3_free(payload);
+    DNSServiceRefDeallocate(sref);
+    return;
+  }
+
+  payload->sref = sref;
+  payload->port = port;
+
+  char* who = strtok(name, ".");
+
+  if (who == NULL) {
+    u3l_log("mdns: invalid domain name");
+    c3_free(payload);
+    DNSServiceRefDeallocate(sref);
+    return;
+  }
+
+  char* who2 = c3_malloc(strlen(who) + 1);
+  snprintf(who2, sizeof who2, "~%s", who);
+  payload->who = who2;
+
+  struct addrinfo hints;
+  memset(&hints, 0, sizeof(hints));
+  hints.ai_family = AF_INET; // Request only IPv4 addresses
+  hints.ai_socktype = SOCK_STREAM; // TCP socket
+
+  uv_getaddrinfo_t* req = (uv_getaddrinfo_t*)c3_malloc(sizeof(uv_getaddrinfo_t));
+  req->data = (void*)payload;
+
+  uv_loop_t* loop = uv_default_loop();
+  int error = uv_getaddrinfo(loop, req, getaddrinfo_cb, host, NULL, &hints);
+
+  if (error < 0) {
+    u3l_log("mdns: getaddrinfo error: %s\n", uv_strerror(error));
+    c3_free(payload);
+    c3_free(req);
+    DNSServiceRefDeallocate(sref);
+  }
+}
+
+static void poll_cb(uv_poll_t *handle, int status, int events) {
+  DNSServiceRef sref = (DNSServiceRef) handle->data;
+  int err = DNSServiceProcessResult(sref);
+}
+
+static void dns_sd_cb(DNSServiceRef sref, mdns_payload* payload) {
+  int fd = DNSServiceRefSockFD(sref);
+  uv_loop_t* loop = uv_default_loop();
+  uv_poll_t* poll = (uv_poll_t *)c3_malloc(sizeof(uv_poll_t));
+  poll->data = (void *)sref;
+  payload->poll = poll;
+  uv_poll_init(loop, poll, fd);
+  uv_poll_start(poll, UV_READABLE, poll_cb);
```

Comment:
If `poll` has the same lifetime as `payload` (see above)`, you can remove the malloc here and just `uv_poll_init(loop, &(payload->poll), fd);`

Reply:
And done.